crowdfunding/users/views.py

```diff
@@ -0,0 +1,89 @@
+from django.http import Http404
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import status, generics, permissions
+from rest_framework.permissions import IsAuthenticated   
+
+
+from .models import CustomUser
+from .serializers import CustomUserSerializer, ChangePasswordSerializer, CustomUserDetailSerializer
+from projects.permissions import IsUserOrBanished
+
+class CustomUserList(APIView):
+
+    def get(self, request):
+        users = CustomUser.objects.all()
+        serializer = CustomUserSerializer(users, many=True)
+        return Response(serializer.data)
+
+    def post(self, request):
+        serializer = CustomUserSerializer(data=request.data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data)
+        return Response(serializer.errors)
+
+class CustomUserDetail(generics.RetrieveAPIView):
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly, IsUserOrBanished]
+    queryset = CustomUser.objects.all()
+    serializer_class = CustomUserDetailSerializer
+
+    
+    
+    # def get_object(self, pk):
+    #     try:
+    #         return CustomUser.objects.get(pk=pk)
+    #     except CustomUser.DoesNotExist:
+    #         raise Http404
+    
+    # def get(self, request, pk):
+    #     user = self.get_object(pk)
+    #     serializer = CustomUserDetailSerializer(user)
+    #     return Response(serializer.data)
+
+# class CustomUserDetail(APIView):
+
+#     def get_object(self, pk):
+#         try:
+#             return CustomUser.objects.get(pk=pk)
+#         except CustomUser.DoesNotExist:
+#             raise Http404
+    
+#     def get(self, request, pk):
+#         user = self.get_object(pk)
+#         serializer = CustomUserSerializer(user)
+#         return Response(serializer.data)
+
+class ChangePasswordView(generics.UpdateAPIView):
+    """
+    An endpoint for changing password.
+    """
+    serializer_class = ChangePasswordSerializer
+    model = CustomUser
+    permission_classes = (IsAuthenticated,)
+
+    def get_object(self, queryset=None):
+        obj = self.request.user
+        return obj
+
+    def update(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        serializer = self.get_serializer(data=request.data)
+
+        if serializer.is_valid():
+            # Check old password
+            if not self.object.check_password(serializer.data.get("old_password")):
+                return Response({"old_password": ["You gnome nothing John Snow - password is incorrect, try again"]}, status=status.HTTP_400_BAD_REQUEST)
+            # set_password also hashes the password that the user will get
+            self.object.set_password(serializer.data.get("new_password"))
+            self.object.save()
```

Comment:
i would have put a lot of this logic in the serializer, however this seems to be working just fine!